code/modules/station_goals/bluespace_tap.dm

```diff
@@ -0,0 +1,380 @@
+				input_level = 0	//emergency shutdown unless we're sabotaged
+				desired_level = 0
+			for(var/i = 1, i <= rand(1, 3), i++)
+				var/turf/location = locate(x + rand(-5,5), y + rand(-5,5), z)
+				new /obj/structure/spawner/nether/bluespace_tap(location)
+
+
+
+/obj/machinery/power/bluespace_tap/tgui_data(mob/user)
+	var/data[0]
+
+	data["desiredLevel"] = desired_level
+	data["inputLevel"] = input_level
+	data["points"] = points
+	data["totalPoints"] = total_points
+	data["powerUse"] = actual_power_usage
+	data["availablePower"] = surplus()
+	data["maxLevel"] = max_level
+	data["emagged"] = emagged
+	data["safeLevels"] = safe_levels
+	data["nextLevelPower"] = get_power_use(input_level + 1)
+
+	var/list/listed_items = list()//a list of lists, each inner list equals a datum
+	for(var/key = 1 to length(product_list))
+		var/datum/data/bluespace_tap_product/A = product_list[key]
+		listed_items.Add(list(list(	//note that adding a list to a list in BYOND adds the indiviudal list elements
+				"key" = key,		//so to get a list of lists the inner lists need to be wrapped in another list
+				"name" = A.product_name,
+				"price" = A.product_cost)))
+	data["product"] = listed_items
+	return data
+
+
+/obj/machinery/power/bluespace_tap/attack_hand(mob/user)
+	add_fingerprint(user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ghost(mob/user)
+	tgui_interact(user)
+
+/obj/machinery/power/bluespace_tap/attack_ai(mob/user)
+	tgui_interact(user)
+
+//produces and vends the product with the desired key
+/obj/machinery/power/bluespace_tap/proc/produce(key)
+	var/datum/data/bluespace_tap_product/A = product_list[key]
+	if(!A)
+		return
+	if(A.product_cost > points)
+		return
+	points -= A.product_cost
+	A.product_cost = round(1.2 * A.product_cost, 1)
+	playsound(src, 'sound/magic/blink.ogg', 50)
+	do_sparks(2, FALSE, src)
+	new A.product_path(get_turf(src))
+
+
+
+//UI stuff below
+
+/obj/machinery/power/bluespace_tap/tgui_act(action, params)
+	if(..())
+		return 1
+	switch(action)
+		if("decrease")
+			decrease_level()
+		if("increase")
+			increase_level()
+		if("set")
+			set_level(text2num(params["set_level"]))
+		if("vend")//it's not really vending as producing, but eh
+			var/key = text2num(params["target"])
+			produce(key)
+
+/obj/machinery/power/bluespace_tap/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_default_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "BluespaceTap", name, 650, 400, master_ui, state)	//Size of window tbd
+		ui.open()
+
+//emaging provides slightly more points but at much greater risk
+/obj/machinery/power/bluespace_tap/emag_act(mob/living/user as mob)
+	if(!emagged)
+		emagged = TRUE
+		do_sparks(5, FALSE, src)
+		if(user)
+			user.visible_message("[user] emags [src].","<span class='warning'>You override the safety protocols.</span>")
```

Comment:
Need spans here. Also the text `X emags Y` is a bit short. It can use some more flavor I'd say